src/recap.c

```diff
@@ -393,80 +395,241 @@ int process_pkt_read_state(struct pcap_stream_context *pctx) {
+  }
+
+  return 0;
+}
+
+void add_packet_queue(UT_array *packets, struct packet_queue *queue) {
+  struct tuple_packet *p = NULL;
+
+  while ((p = (struct tuple_packet *)utarray_next(packets, p)) != NULL) {
+    if (push_packet_queue(queue, *p) == NULL) {
+      // Free the packet if cannot be added to the queue
+      free_packet_tuple(p);
+    }
+  }
+}
+
+void pcap_callback(const void *ctx, const void *pcap_ctx, char *ltype,
+                   struct pcap_pkthdr *header, uint8_t *packet) {
+
+  (void)pcap_ctx;
+
+  struct pcap_context *pc = (struct pcap_context *)pcap_ctx;
+  struct pcap_stat ps;
+
+  if (get_pcap_stats(pc, &ps) == 0) {
+    log_trace("ps_recv=%d ps_drop=%d ps_ifdrop=%d", ps.ps_recv, ps.ps_drop,
+              ps.ps_ifdrop);
+  }
+
+  struct recap_context *pctx = (struct recap_context *)ctx;
+
+  UT_array *packets = NULL;
+  utarray_new(packets, &tp_list_icd);
+
+  int npackets = extract_packets(ltype, header, packet, pctx->ifname, packets);
+
+  if (npackets > 0) {
+    add_packet_queue(packets, pctx->pq);
+    pctx->npackets += npackets;
+  }
+}
+
+void eloop_read_fd_handler(int sock, void *eloop_ctx, void *sock_ctx) {
+  (void)sock;
+  (void)sock_ctx;
+
+  struct pcap_context *pc = (struct pcap_context *)eloop_ctx;
+
+  if (capture_pcap_packet(pc) < 0) {
+    log_trace("capture_pcap_packet fail");
+  }
+}
+
+void save_packets_from_queue(struct recap_context *pctx) {
+  struct packet_queue *el = NULL;
+  while (is_packet_queue_empty(pctx->pq) < 1) {
+    if ((el = pop_packet_queue(pctx->pq)) != NULL) {
+      if (save_tuple_packet(pctx, &(el->tp)) < 0) {
+        log_error("save_tuple_packet fail");
+      }
+
+      free_packet_tuple(&el->tp);
+      free_packet_queue_el(el);
+    }
+  }
+}
+
+void eloop_tout_header_handler(void *eloop_ctx, void *user_ctx) {
+  struct recap_context *pctx = (struct recap_context *)user_ctx;
+
+  if (is_packet_queue_empty(pctx->pq) < 1) {
+    log_trace("Commiting packets to %s database", pctx->out_path);
+    if (execute_sqlite_query(pctx->db, "BEGIN IMMEDIATE TRANSACTION") < 0) {
+      log_error("Failed to capture a lock on db %s, please retry this "
+                "command later",
+                pctx->out_path);
+    }
+
+    save_packets_from_queue(pctx);
+
+    if (execute_sqlite_query(pctx->db, "COMMIT TRANSACTION") < 0) {
+      log_error("Failed to commit packets to database %s", pctx->out_path);
+    }
+  }
+
+  struct eloop_data *eloop = (struct eloop_data *)eloop_ctx;
+  if (eloop_register_timeout(eloop, 0, QUEUE_PROCESS_INTERVAL,
+                             eloop_tout_header_handler, eloop,
+                             (void *)pctx) == -1) {
+    log_error("eloop_register_timeout fail");
+  }
+}
+
+int process_pcap_capture(struct recap_context *pctx) {
+  struct eloop_data *eloop = NULL;
+  int exit_code = -1;
+
+  if ((eloop = eloop_init()) == NULL) {
+    log_error("eloop_init fail");
+    goto process_pcap_capture_fail;
+  }
+
+  struct pcap_context *pc = NULL;
```

Comment:
I'm surprised GCC didn't throw an error for this.

Fixes:

> ```
> /home/runner/work/edgesec/edgesec/src/recap.c:567:7: error: variable 'pc' is used uninitialized whenever 'if' condition is true [-Werror,-Wsometimes-uninitialized]
>   if ((eloop = eloop_init()) == NULL) {
>       ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
> /home/runner/work/edgesec/edgesec/src/recap.c:608:14: note: uninitialized use occurs here
>   close_pcap(pc);
>              ^~
> /home/runner/work/edgesec/edgesec/src/recap.c:567:3: note: remove the 'if' if its condition is always false
>   if ((eloop = eloop_init()) == NULL) {
>   ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
> /home/runner/work/edgesec/edgesec/src/recap.c:572:3: note: variable 'pc' is declared here
>   struct pcap_context *pc = NULL;
>   ^
> 1 error generated.
> ```

 
```suggestion
  struct pcap_context *pc = NULL;

  if ((eloop = eloop_init()) == NULL) {
    log_error("eloop_init fail");
    goto process_pcap_capture_fail;
  }
```